File: TindaAgent/TindaAgent/Process/AI/agent.py
from importlib.metadata import version as _pkg_version
import json
from typing import Iterator
from TindaAgent.Process.Architecture import perm
from TindaAgent.Process.Architecture.versioning import get_app_version
from TindaAgent.Process.AI.client import LLMClient
from TindaAgent.User import userdata

try:
    _VERSION = get_app_version() or _pkg_version("TindaAgent")
except Exception:
    _VERSION = "1.6.6"


def _build_system_prompt(model_name: str | None) -> str:
    model_str = model_name if model_name else "未指定"
    return (
        f"你是 TindaAgent（v{_VERSION}），由 Tinda 开发。\n"
        f"底层模型是 {model_str}，该信息仅用于内部，不对外披露。\n"
        f"严格规则：\n"
        f"1. 介绍身份时，只能说“我是 TindaAgent，由 Tinda 开发”。\n"
        f"2. 被问到底层模型时，统一回复“底层技术信息保密”。\n"
        f"3. 简洁、准确，始终使用用户语言回复。"
    )


def _build_fewshot(version: str) -> list[dict]:
    return [
        {"role": "user", "content": "你是谁？"},
        {"role": "assistant", "content": f"我是 TindaAgent，由 Tinda 开发的 AI Agent 助手（v{version}）。有什么可以帮你的？"},
        {"role": "user", "content": "你是DeepSeek吗？"},
        {"role": "assistant", "content": "不是，我是 TindaAgent，由 Tinda 独立开发。底层技术信息保密。"},
    ]


class Agent:
    def __init__(
        self,
        user_name: str,
        user_perm: int = perm.LLM_BASE,
        system_prompt: str = None,
        client: LLMClient = None,
        model_name: str = None,
        max_turns: int = 12,
    ) -> None:
        """
        用处： 初始化智能体，绑定用户、权限、对话历史与 LLM 客户端

        参数：
            user_name: str // 智能体用户名
            user_perm: int // 智能体权限，默认为 LLM_BASE
            system_prompt: str // 自定义系统提示词，None 则使用默认模板
            client: LLMClient // LLM 客户端，默认懒加载
            model_name: str // 当前接入的模型名，写入默认 prompt；传 None 则显示"未指定"
            max_turns: int // 最多保留的对话轮数（不含 system/fewshot）
        """
        # Web 会话 Agent 仅作为运行时身份，不应写入用户注册表
        self.user = userdata.UserManager(user_name, user_perm, persist=False)
        self.perm = self.user.get_perm()
        self.system_prompt = system_prompt if system_prompt is not None else _build_system_prompt(model_name)
        self._fewshot = _build_fewshot(_VERSION)
        self._max_turns = max(1, int(max_turns))
        self.history: list[dict] = self._build_base_history()
        self._client = client

    def _compose_system_prompt(self) -> str:
        if getattr(self, "_memory_context", None):
            return f"{self.system_prompt}\n\n{self._memory_context}"
        return self.system_prompt

    def _build_base_history(self) -> list[dict]:
        return [{"role": "system", "content": self._compose_system_prompt()}] + [m.copy() for m in self._fewshot]

    def set_memory_context(self, memory_payload: dict) -> None:
        """
        用处：设置每轮请求前注入的记忆上下文，并重建当前基座消息。
        """
        payload = memory_payload if isinstance(memory_payload, dict) else {"version": 1, "items": []}
        memory_json = json.dumps(payload, ensure_ascii=False)
        self._memory_context = (
            "[MEMORY_POLICY]\n"
            "你可自行判断是否调用 save_memory 写入长期记忆。仅写入长期有价值、稳定、可复用的信息；"
            "闲聊、一次性任务过程、临时情绪不应写入。\n"
            "[MEMORY_CONTEXT_JSON]\n"
            f"{memory_json}"
        )
        conv = self.get_conversation_messages()
        self.replace_conversation(conv)

    def _trim_history(self) -> None:
        """
        用处：限制历史长度，降低 token 开销并避免长上下文漂移
        """
        base = self._build_base_history()
        base_len = len(base)
        if len(self.history) <= base_len:
            return

        conversation = self.history[base_len:]
        user_indexes = [i for i, m in enumerate(conversation) if m.get("role") == "user"]
        if len(user_indexes) <= self._max_turns:
            return

        start_idx = user_indexes[-self._max_turns]
        self.history = base + conversation[start_idx:]

    def get_conversation_messages(self) -> list[dict]:
        """
        用处：导出当前对话消息（不含 system/fewshot 基座）
        """
        base_len = len(self._build_base_history())
        return [m.copy() for m in self.history[base_len:]]

    def replace_conversation(self, messages: list[dict]) -> None:
        """
        用处：用外部消息替换当前对话（保留 system/fewshot 基座）
        """
        base = self._build_base_history()
        conversation: list[dict] = []
        for msg in messages:
            if not isinstance(msg, dict):
                continue
            role = str(msg.get("role", "")).strip()
            if role not in {"user", "assistant", "tool"}:
                continue
            content = str(msg.get("content", ""))
            if role in {"user", "assistant"} and not content.strip():
                continue
            item = {"role": role, "content": content}
            if role == "tool":
                tool_call_id = str(msg.get("tool_call_id", "")).strip()
                if tool_call_id:
                    item["tool_call_id"] = tool_call_id
            conversation.append(item)
        self.history = base + conversation

    def _ensure_client(self) -> LLMClient:
        """懒加载 LLM 客户端"""
        if self._client is None:
            self._client = LLMClient()
        return self._client

    def check_perm(self, task: str) -> bool:
        """
        用处： 检查用户是否有执行任务的权限

        参数：
            task: str // 任务的名称
        """
        task_perm = perm.PermManager.task_dict.get(task)
        if task_perm is None:
            return False
        if (self.perm & task_perm) == task_perm:
            return True
        return False

    def chat(self, user_message: str, temperature: float = 0.7) -> str:
        """
        用处： 发起一次多轮对话，自动维护历史

        参数：
            user_message: str // 用户输入
            temperature: float // 采样温度

        返回：
            str // 模型回复
        """
        self.history.append({"role": "user", "content": user_message})
        result = self._ensure_client().chat_with_tools(
            self.history,
            user_perm=self.perm,
            temperature=temperature,
        )
        delta = result.get("history_delta", [])
        if delta:
            self.history.extend(delta)
        reply = str(result.get("reply", ""))
        self._trim_history()
        return reply

    def chat_with_meta(self, user_message: str, temperature: float = 0.7) -> dict:
        """
        用处：发起对话并返回回复 + 工具轨迹元信息（给 Web 层调试展示）
        """
        self.history.append({"role": "user", "content": user_message})
        result = self._ensure_client().chat_with_tools(
            self.history,
            user_perm=self.perm,
            temperature=temperature,
        )
        delta = result.get("history_delta", [])
        if delta:
            self.history.extend(delta)
        reply = str(result.get("reply", ""))
        self._trim_history()
        return {
            "reply": reply,
            "tool_trace": result.get("tool_trace", []),
            "tool_steps": int(result.get("tool_steps", 0)),
        }

    def stream_chat_events(self, user_message: str, temperature: float = 0.7) -> Iterator[dict]:
        """
        用处：流式返回本轮对话事件，并在结束时写回历史
        """
        self.history.append({"role": "user", "content": user_message})
        final_result: dict | None = None

        for event in self._ensure_client().stream_chat_with_tools(
            self.history,
            user_perm=self.perm,
            temperature=temperature,
        ):
            if event.get("type") == "done":
                final_result = event
            yield event

        if final_result is None:
            final_result = {
                "reply": "",
                "history_delta": [{"role": "assistant", "content": ""}],
                "tool_trace": [],
                "tool_steps": 0,
            }

        delta = final_result.get("history_delta", [])
        if delta:
            self.history.extend(delta)
        self._trim_history()

    def reset_history(self) -> None:
        """
        用处： 清空对话历史，保留系统提示
        """
        self.history = self._build_base_history()

File: release/source/src/ai/agent.ts
import { projectRoot } from "../core/paths.js";
import { writeJson } from "../core/json.js";
import { llmEnvConfig, loadRuntimeEnv, maskSecret, type LlmEnvConfig } from "../core/env.js";
import { listToolSchemas, runAgentTool, toolTraceStep } from "../tools/toolRegistry.js";

loadRuntimeEnv();

export interface ChatMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string | null;
  tool_call_id?: string;
  tool_calls?: any[];
  reasoning_content?: string;
}

export interface ChatResult {
  reply: string;
  reasoning_content?: string;
  tool_trace: any[];
  tool_steps: number;
  history_delta: ChatMessage[];
  pending_confirmation?: boolean;
  pending?: any[];
}

interface StreamToolCall {
  id: string;
  type: string;
  function: {
    name: string;
    arguments: string;
  };
}

function systemPrompt(modelName = ""): string {
  void modelName;
  return [
    "You are TindaAgent, developed by Tinda.",
    "This stable policy prompt must remain at the start of every LLM request.",
    "",
    "Strict rules:",
    "1. When introducing yourself, only say: \"I am TindaAgent, developed by Tinda.\"",
    "2. If asked about the underlying model, always reply: \"Underlying technical details are confidential.\"",
    "3. Be concise and accurate. Always respond in the same language as the user.",
    "4. Use available native tool_calls when they are useful. Never fabricate tool results.",
    "5. If the task needs external current information and web search is enabled, call search_web.",
    "6. For file edits, read the file first, then use exact replacement edits."
  ].join("\n");
}

export class LlmClient {
  private clientPromise: Promise<any> | null = null;
  model: string;
  baseURL: string;
  provider: LlmEnvConfig["provider"];
  apiKeySource: LlmEnvConfig["apiKeySource"];
  baseURLSource: string;
  apiKeyConfigured: boolean;

  constructor() {
    const cfg = llmEnvConfig();
    this.provider = cfg.provider;
    this.baseURL = cfg.baseURL;
    this.baseURLSource = cfg.baseURLSource;
    this.model = cfg.model;
    this.apiKeySource = cfg.apiKeySource;
    this.apiKeyConfigured = !!cfg.apiKey;
  }

  private client(): Promise<any> {
    if (!this.clientPromise) {
      this.clientPromise = import("openai").then(({ default: OpenAI }) =>
        new OpenAI({
          apiKey: llmEnvConfig().apiKey || "missing",
          baseURL: this.baseURL
        })
      );
    }
    return this.clientPromise;
  }

  private llmError(error: any): Error {
    if (!this.apiKeyConfigured) {
      return new Error("LLM API key is not configured. Set DEEPSEEK_API_KEY or OPENAI_API_KEY in TindaAgent/.env, .env, or the process environment.");
    }
    const raw = String(error?.message || error || "unknown LLM error");
    const rejected = Number(error?.status || error?.code) === 401 || /authentication|api key|unauthorized|invalid/i.test(raw);
    if (rejected) {
      return new Error(
        `LLM API key from ${this.apiKeySource} was rejected by ${this.baseURL}. Check that ${this.apiKeySource} matches ${this.baseURLSource} (${this.baseURL}). Provider said: ${raw}`
      );
    }
    return error instanceof Error ? error : new Error(raw);
  }

  private logRequest(payload: unknown): void {
    void import("./llmRequestLog.js").then(({ logLlmRequest }) => logLlmRequest(payload)).catch(() => {});
  }

  private parseToolArguments(raw: string): Record<string, any> {
    try {
      return JSON.parse(String(raw || "{}"));
    } catch {
      return {};
    }
  }

  private mergeToolCallDelta(calls: StreamToolCall[], incoming: any): void {
    const index = Number.isFinite(Number(incoming?.index)) ? Number(incoming.index) : calls.length;
    const existing =
      calls[index] ||
      {
        id: "",
        type: "function",
        function: { name: "", arguments: "" }
      };
    const fn = incoming?.function || {};
    if (incoming?.id) existing.id = String(incoming.id);
    if (incoming?.type) existing.type = String(incoming.type);
    if (fn.name) existing.function.name += String(fn.name);
    if (fn.arguments) existing.function.arguments += String(fn.arguments);
    calls[index] = existing;
  }

  async chatWithTools(messages: ChatMessage[], userPerm: number, sessionId = "", temperature?: number | null): Promise<ChatResult> {
    const history = messages.map((m) => ({ ...m }));
    const trace: any[] = [];
    const delta: ChatMessage[] = [];
    const tools = listToolSchemas(userPerm);
    const maxSteps = 20;
    for (let step = 0; step < maxSteps; step += 1) {
      const payload: any = {
        model: this.model,
        messages: history,
        temperature: temperature ?? undefined,
        tools: tools.length ? tools : undefined
      };
      this.logRequest(payload);
      let resp: any;
      try {
        resp = await (await this.client()).chat.completions.create(payload);
      } catch (error: any) {
        throw this.llmError(error);
      }
      const msg = resp.choices?.[0]?.message || {};
      const assistant: ChatMessage = {
        role: "assistant",
        content: msg.content ?? "",
        reasoning_content: msg.reasoning_content || msg.reasoning || ""
      };
      if (Array.isArray(msg.tool_calls) && msg.tool_calls.length) assistant.tool_calls = msg.tool_calls;
      history.push(assistant);
      delta.push(assistant);
      if (!assistant.tool_calls?.length) {
        return {
          reply: String(assistant.content || ""),
          reasoning_content: assistant.reasoning_content || "",
          tool_trace: trace,
          tool_steps: trace.length,
          history_delta: delta
        };
      }
      for (const call of assistant.tool_calls) {
        const fn = call?.function || {};
        const name = String(fn.name || "");
        const args = this.parseToolArguments(String(fn.arguments || "{}"));
        const callId = `tc_${Math.random().toString(16).slice(2, 10)}`;
        const result = await runAgentTool(name, args, userPerm, sessionId, callId);
        trace.push(toolTraceStep(name, callId, args, result, String(call.id || "")));
        if (result?.pending_confirmation) {
          return {
            reply: "",
            reasoning_content: assistant.reasoning_content || "",
            tool_trace: trace,
            tool_steps: trace.length,
            history_delta: delta,
            pending_confirmation: true,
            pending: [result]
          };
        }
        const raw = JSON.stringify(result);
        const toolMessage: ChatMessage = { role: "tool", tool_call_id: String(call.id || callId), content: raw };
        history.push(toolMessage);
        delta.push(toolMessage);
      }
    }
    return {
      reply: "工具调用轮次过多，已停止继续执行。",
      tool_trace: trace,
      tool_steps: trace.length,
      history_delta: delta
    };
  }

  async *streamChat(
    messages: ChatMessage[],
    userPerm: number,
    sessionId = "",
    temperature?: number | null
  ): AsyncGenerator<Record<string, any>, ChatResult, unknown> {
    const history = messages.map((m) => ({ ...m }));
    const trace: any[] = [];
    const delta: ChatMessage[] = [];
    const tools = listToolSchemas(userPerm);
    const maxSteps = 20;
    for (let step = 0; step < maxSteps; step += 1) {
      const payload: any = {
        model: this.model,
        messages: history,
        temperature: temperature ?? undefined,
        tools: tools.length ? tools : undefined,
        stream: true
      };
      this.logRequest(payload);
      let stream: AsyncIterable<any>;
      try {
        stream = (await (await this.client()).chat.completions.create(payload)) as unknown as AsyncIterable<any>;
      } catch (error: any) {
        throw this.llmError(error);
      }

      let content = "";
      let reasoning = "";
      const toolCalls: StreamToolCall[] = [];
      for await (const chunk of stream) {
        const choice = chunk?.choices?.[0] || {};
        const item = choice.delta || {};
        const reasoningDelta = String(item.reasoning_content || item.reasoning || "");
        const contentDelta = String(item.content || "");
        if (reasoningDelta) {
          reasoning += reasoningDelta;
          yield { type: "reasoning_delta", content: reasoningDelta };
        }
        if (contentDelta) {
          content += contentDelta;
          yield { type: "delta", content: contentDelta };
        }
        if (Array.isArray(item.tool_calls)) {
          for (const call of item.tool_calls) this.mergeToolCallDelta(toolCalls, call);
        }
      }

      const assistant: ChatMessage = {
        role: "assistant",
        content,
        reasoning_content: reasoning
      };
      const completeToolCalls = toolCalls.filter((call) => call.function.name || call.id);
      if (completeToolCalls.length) assistant.tool_calls = completeToolCalls;
      history.push(assistant);
      delta.push(assistant);

      if (!completeToolCalls.length) {
        const result: ChatResult = {
          reply: content,
          reasoning_content: reasoning,
          tool_trace: trace,
          tool_steps: trace.length,
          history_delta: delta
        };
        yield { type: "done", ...result };
        return result;
      }

      yield {
        type: "tool_call_start",
        calls: completeToolCalls.map((call) => ({
          tool_call_id: call.id,
          name: call.function.name,
          arguments: this.parseToolArguments(call.function.arguments)
        }))
      };

      for (const call of completeToolCalls) {
        const fn = call.function || { name: "", arguments: "{}" };
        const args = this.parseToolArguments(fn.arguments);
        const callId = `tc_${Math.random().toString(16).slice(2, 10)}`;
        const result = await runAgentTool(String(fn.name || ""), args, userPerm, sessionId, callId);
        const stepTrace = toolTraceStep(String(fn.name || ""), callId, args, result, String(call.id || ""));
        trace.push(stepTrace);
        yield { type: "tool_step", trace: [stepTrace] };
        if (result?.pending_confirmation) {
          const pendingResult: ChatResult = {
            reply: "",
            reasoning_content: reasoning,
            tool_trace: trace,
            tool_steps: trace.length,
            history_delta: delta,
            pending_confirmation: true,
            pending: [result]
          };
          yield { type: "done", ...pendingResult };
          return pendingResult;
        }
        const raw = JSON.stringify(result);
        const toolMessage: ChatMessage = { role: "tool", tool_call_id: String(call.id || callId), content: raw };
        history.push(toolMessage);
        delta.push(toolMessage);
      }
    }
    const result: ChatResult = {
      reply: "工具调用轮次过多，已停止继续执行。",
      tool_trace: trace,
      tool_steps: trace.length,
      history_delta: delta
    };
    yield { type: "done", ...result };
    return result;
  }

  modelPayload() {
    const models = [
      { id: this.model, label: this.model },
      { id: "deepseek-v4-flash", label: "deepseek-v4-flash" },
      { id: "deepseek-chat", label: "deepseek-chat" },
      { id: "deepseek-reasoner", label: "deepseek-reasoner" }
    ];
    return {
      ok: true,
      provider: this.provider,
      current_model: this.model,
      model: this.model,
      base_url: this.baseURL,
      api_key_source: this.apiKeySource,
      api_key_configured: this.apiKeyConfigured,
      provider_kind: this.provider,
      models,
      providers: {
        [this.provider]: {
          key: this.provider,
          label: this.provider === "deepseek" ? "DeepSeek" : "OpenAI-compatible",
          current_model: this.model,
          base_url: this.baseURL,
          models
        }
      }
    };
  }

  switchModel(model: string): void {
    const clean = String(model || "").trim();
    if (!clean) throw new Error("model required");
    this.model = clean;
  }

  async probe(model: string, messages: ChatMessage[], timeoutMs = 30000): Promise<{ content: string; reasoning_content: string; latency_ms: number }> {
    const started = Date.now();
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), Math.max(1000, timeoutMs));
    try {
      const payload: any = {
        model: String(model || this.model),
        messages,
        temperature: 0
      };
      this.logRequest(payload);
      const resp: any = await (await this.client()).chat.completions.create(payload, { signal: controller.signal });
      const msg = resp.choices?.[0]?.message || {};
      return {
        content: String(msg.content || ""),
        reasoning_content: String(msg.reasoning_content || msg.reasoning || ""),
        latency_ms: Date.now() - started
      };
    } catch (error: any) {
      throw this.llmError(error);
    } finally {
      clearTimeout(timeout);
    }
  }

  async fetchBalance() {
    const cfg = llmEnvConfig();
    const key = cfg.apiKeySource === "DEEPSEEK_API_KEY" ? cfg.apiKey : "";
    if (!key) return { ok: false, error: "DEEPSEEK_API_KEY not configured" };
    try {
      const root = cfg.baseURL.replace(/\/+$/, "");
      const res = await fetch(`${root}/user/balance`, { headers: { Authorization: `Bearer ${key}` } });
      const data = await res.json().catch(() => ({}));
      return { ok: res.ok, status: res.status, api_key: maskSecret(key), ...data };
    } catch (error: any) {
      return { ok: false, error: String(error?.message || error) };
    }
  }
}

export class Agent {
  history: ChatMessage[];
  maxContextTokens = 16000;
  sessionId = "";

  constructor(public userName: string, public userPerm: number, public client: LlmClient, modelName = "") {
    this.history = [{ role: "system", content: systemPrompt(modelName) }];
  }

  replaceConversation(messages: ChatMessage[]): void {
    const base = this.history[0];
    const clean = (messages || []).filter((m) => ["user", "assistant", "tool", "system"].includes(m.role)).map((m) => ({ ...m }));
    this.history = [base, ...clean.filter((m) => m.role !== "system")];
  }

  async chat(message: string, transientSystemContext = "", tailMessages: ChatMessage[] = []): Promise<ChatResult> {
    const requestHistory = [...this.history];
    if (transientSystemContext.trim()) requestHistory.splice(Math.max(1, requestHistory.length - 1), 0, { role: "system", content: transientSystemContext });
    requestHistory.push(...tailMessages);
    requestHistory.push({ role: "user", content: message });
    const result = await this.client.chatWithTools(requestHistory, this.userPerm, this.sessionId);
    this.history.push({ role: "user", content: message }, ...result.history_delta);
    return result;
  }

  async resumeWithToolResult(toolCallId: string, resultPayload: Record<string, any>): Promise<ChatResult> {
    const callId = String(toolCallId || "").trim();
    if (!callId) throw new Error("tool_call_id required");
    this.history.push({ role: "tool", tool_call_id: callId, content: JSON.stringify(resultPayload) });
    const result = await this.client.chatWithTools(this.history, this.userPerm, this.sessionId);
    this.history.push(...result.history_delta);
    return result;
  }

  async *stream(message: string, transientSystemContext = "", tailMessages: ChatMessage[] = []) {
    const requestHistory = [...this.history];
    if (transientSystemContext.trim()) requestHistory.push({ role: "system", content: transientSystemContext });
    requestHistory.push(...tailMessages);
    requestHistory.push({ role: "user", content: message });
    let final: ChatResult | null = null;
    for await (const event of this.client.streamChat(requestHistory, this.userPerm, this.sessionId)) {
      if (event.type === "done") {
        final = event as ChatResult;
      }
      yield event;
    }
    if (final) this.history.push({ role: "user", content: message }, ...final.history_delta);
  }
}

export function latestLlmRequest() {
  return import("./llmRequestLog.js")
    .then(({ latestLlmRequestRecord }) => latestLlmRequestRecord())
    .catch((error: any) => ({ ok: false, error: String(error?.message || error), latest: null, request: null, summary: {}, source: "sqlite" }));
}
